src/jsonstream.cpp

```diff
@@ -0,0 +1,201 @@
+//
+// Created by jpiotte on 10/23/19.
+//
+
+#include <simdjson/isadetection.h>
+#include <atomic>
+#include <simdjson/padded_string.h>
+#include "simdjson/jsonstream.h"
+
+using namespace simdjson;
+
+//Architecture find_best_supported_implementation();
+//int json_parse_implementation(ParsedJson &pj, bool realloc_if_needed);
+//
+//
+//// function pointer type for json_parse
+//using json_parse_functype = int(const char *buf, size_t len, size_t batch_size, ParsedJson &pj, bool realloc_if_needed);
+//
+//// Pointer that holds the json_parse implementation corresponding to the
+//// available SIMD instruction set
+//extern std::atomic<json_parse_functype *> json_parse_ptr;
+//
+
+
+
+
+
+
+JsonStream::JsonStream(const char *buf, size_t len, size_t batch_size) {
+//    this->best_implementation = find_best_supported_implementation();
+    this->buf = buf;
+    this->len = len;
+    this->batch_size = batch_size;
+}
+
+JsonStream::~JsonStream() {
+    batch_size = 0;
+    len = 0;
+    batch_size = 0;
+    next_json = 0;
+    current_buffer_loc = 0;
+    n_parsed_docs = 0;
+    error_on_last_attempt= false;
+    load_next_batch = true;
+}
+
+
+int JsonStream::json_parse(ParsedJson &pj, bool realloc_if_needed) {
+    //return json_parse_ptr.load(std::memory_order_relaxed)(buf, len, batch_size, pj, realloc_if_needed);
+
+    if (pj.byte_capacity == 0) {
+        bool allocok = pj.allocate_capacity(batch_size, batch_size);
+        if (!allocok) {
+            std::cerr << "can't allocate memory" << std::endl;
+            return false;
+        }
+    }
+    else if (pj.byte_capacity < batch_size) {
+        return simdjson::CAPACITY;
+    }
+
+    //Quick heuristic to see if it's worth parsing the remaining data in the batch
+    if(!load_next_batch && n_bytes_parsed > 0) {
+        auto remaining_data = batch_size - current_buffer_loc;
+        auto avg_doc_len = (float) n_bytes_parsed / n_parsed_docs;
```

Comment:
Minor thing perhaps, but I'd `const` both these. Same for `bool allocok` above and `stage1_is_ok` below and `int res` below.

Reply:
Agreed.